ZEPXXX-priority-election.md

```diff
@@ -0,0 +1,231 @@
+
+Consider the following scenario:  
+NodeA : {nodePriority = 3, targetPriority = 3}  
+NodeB : {nodePriority = 2, targetPriority = 3}  
+NodeC : {nodePriority = 1, targetPriority = 3}  
+
+NodeA is the current leader and it dies. `priorityDecay = 1`.
+
+![priority-election](ZEP006/priority_election.jpg)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The priority election is implemented by the follower.
+Previously, on heartbeat timeout a follower immediately starts an election by sending a poll request.
+The heartbeat timeout was also randomized in order to prevent multiple nodes starting the election at the same time.
+With the priority election we do not use a randomized heartbeat timeout.
+Instead we use the configured electionTimeout.
+On heartbeat timeout, the follower implements the following algorithm.
+
+
+```java
+onHeartbeatTimeout() {
+    if(nodePriority >= targetPriority) {
+        sendPollRequest();
+    } else {
+        // reduce priority in the next heartbeatTimeout
+        heartbeatTimer = schedule(electionTimeout, () -> {
+            targetPriority = targetPriority - priorityDecay;
+            onHeartbeatTimeout();
+        })
+    }
+}
+```
+
+
+### Assigning priority
+
+The node with highest priority is called the primary for the partition.
+If a node is primary for two partitions a and b, then the next priority for each partitions should be assigned to different nodes.
+
+Here is an example how we assign the priorities. Each node gets a priority value for each partition.
+
+| node\partition | 1 | 2 | 3 | 4 | 5 | 6 |
+| --------       |:------:| -----:|:-----:| -----:|:------:| -----:|
+| 0              | 3 | 1 | 2 | 3 | 2 | 1 |
+| 1              | 2 | 3 | 1 | 1 | 3 | 2 |
+| 2              | 1 | 2 | 3 | 2 | 1 | 3 |
+
+Node 0 is the primary for partitions 1 and 4. If node 1 has the next priority for both partitions, then when node 0 dies node 1 will become the leader for both partitions.
+This is not optimal as the leadership is not balanced. Hence it is preferable to make different nodes as the secondary priorities. As in the above example, node 1 is the secondary for partition 1, while node 2 is the secondary for partition 4.
+
+Here is how the priority is assigned.
+The primary node get the highest priority which is equal to the replicationFactor.
+
+```java
+primaryMember.priority = replicationFactor;
+if(partition/clustersize  % 2 == 0) {
+    nextPriority = 1;
+    foreach (member : nonPrimaryPartitionMembers) {
+        member.priority = nextPriority;
+        nextPriority++;
+    }
+} else {
+    nextPriority = replicationFactor - 1
+    foreach (member : nonPrimaryPartitionMembers) {
```

Comment:
What is the content of this list?

Reply:
Is it sorted?

Reply:
This should be the members of a replication group without the primary. The members of a replication group is determined at the startup of a cluster. It is pre-determined based on the cluster size, partition count and replication factor.

It is not sorted. But I think it does not matter as long as the nodes use the same order. In my spike, I used the same order in which it is calculated by the `buildPartitions()` method and which is passed to the `PartitionMetadata`. I will verify this and update the document.